### PR TITLE
Deduplicate data_needed

### LINKED ISSUE #6497: Alternatives for current ensure_communicating
We've recently been refactoring the worker state machine (https://github.com/dask/distributed/issues/5736)

The new system works great for single-task-based transitions but feels awkward as soon as we're trying to treat groups or batches of tasks similarly which is particularly important for fetching remote data. We are trying to not perform a network request per key but rather try to perform a request per worker and try to get as many keys of it as reasonable.

This mechanism worked so far by running `Worker.(_)ensure_communicating` frequently which iterates over all tasks and schedules a fetch/gather_dep (pseudo code below)

```python
def ensure_communicating(...):
    skip = []
    while self.data_needed and free_slots_for_fetch:
        ts = self.data_needed.pop() # highest priority task
        worker = pick_good_worker(ts) # not busy, maybe localhost, etc.
        if not worker:
            skip.add(ts)
        more_keys_on_worker = select_keys_for_gather(worker)
        schedule_gather_dep(ts, more_keys_on_worker, worker) # This schedules a coroutine and actually fetches the data
    self.data_needed.update(skip) # data needed is a heap so this is non-linear
```

This has a couple of problems
1. It's unclear when to run this function. Past iterations just ran this frequently (on every incoming request and in various code places where we guess there may be something that needs transitioning
2. We're iterating over all keys, all the time. This is bad for performance although not critical on the worker https://github.com/dask/distributed/pull/6462#issuecomment-1138764541
3. If we're calling this too soon, e.g. within a released->fetch transition, we may get poor scheduling decisions by fetching tasks prematurely since we didn't work off all fetch recommendations before calling this. For this reason, the `EnsureCommunicatingAfterTransitions ` instruction was introduced (see also https://github.com/dask/distributed/pull/6462) to mimic earlier behavior

## Proposal

I think we can get rid of `Worker.data_needed` entirely. Since we're storing everything in heaps, by now, `Worker.data_needed_per_worker` carries everything we need if we change how ensure_communicating works which might have other interesting side effects.

To illustrate this, consider the following two helper functions

```python
def get_most_important_worker_to_fetch_from(self):
    """
    Get the remote worker that holds the _most important_ data we should fetch. This is typically inferred by looking at task priorities.
    This could also be used to modify what we consider most important worker to fetch from, e.g. due to load/business, memory pressure, etc.
    """

    # Linear in num_workers
    res = None
    prio = None
    for w, heap in self.data_needed_per_worker.items():
        nprio = heap.peek().priority
        if w not in self.busy_workers and (res is None or nprio > prio):
            res = w
            prio = nprio
    return res

def get_tasks_for_worker(self, w, limit=42):
    """Get a list of tasks on worker `w` such that combined they have at most `limit` estimated bytes"""
    # This is more or less the current `Worker._select_keys_for_gather`
    # Linear in number of tasks in local state fetch on remote worker
    res = []
    size = 0
    for ts in self.data_needed_per_worker[w]:
        size += ts.get_nbytes()
        res.append(ts)
        if size > limit:
            break
    return res
```

With these two helpers we should be able to construct a new, simplified version if we switch our loop from tasks to workers

```python
def _new_ensure_communicate_with_tasks_by_worker():
    if network_limits_not_reached and worker := self.get_most_important_worker_to_fetch_from():
        tasks_to_fetch = self.get_tasks_for_worker(worker)
        instruction = GatherDep(
            worker,
            tasks_to_fetch,
        )
```

or even change the `GatherDep` instruction to worker-only and delay the decision of what tasks to fetch until this instruction is acted on, which typically only happens after all transitions in `Worker.transitions` are worked off

```python
def _new_ensure_communicate_with_tasks_by_worker():
    if network_limits_not_reached and worker := self.get_most_important_worker_to_fetch_from():
        instruction = GatherDep(worker)

def gather_dep(self, worker):
    tasks_to_fetch = self.get_tasks_for_worker(worker)
    ...
```

cc @crusaderky

### PR DESCRIPTION
**Pull request by crusaderky**
- Partially closes #6497
- Complements #6462 
- Remove `data_needed`; rename `data_needed_per_worker` to `data_needed`.
- Fixes use case described in https://github.com/dask/distributed/pull/6462#issuecomment-1138764541, where calling `_ensure_communicating` a second time in short succession would cost `O(t*log(t))`, where t is the number of tasks exclusively held by workers in flight or busy. The cost is now `O(w)`, where w is the number of workers in flight or busy.

### CODE REVIEW

File: distributed/worker_state_machine.py [hunk 1]
```diff
@@ -1217,54 +1211,78 @@ def _purge_state(self, ts: TaskState) -> None:
...
+                continue
+            if worker in self.in_flight_workers or worker in self.busy_workers:
+                busy[worker] = tasks
+                continue
+
+            to_sort.append(
+                (
+                    get_address_host(worker) != host,  # False < True
+                    tasks.peek().priority,
+                    -len(tasks),
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:1240]: Instead of number of tasks, could've measured total nbytes and/or number of tasks that only exists on this worker.
This is more crude but it has the benefit of simplicity.

File: distributed/worker_state_machine.py [hunk 2]
```diff
@@ -1217,54 +1211,78 @@ def _purge_state(self, ts: TaskState) -> None:
...
+                    random.random(),
+                    worker,
+                    tasks,
+                )
+            )
+
+        self.data_needed.clear()
+        self.data_needed.update(
+            (worker, tasks) for _, _, _, _, worker, tasks in sorted(to_sort)
+        )
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:1250]: Note that, if your run `_ensure_communicating` twice in a row, on the second round `to_sort` will be empty

File: distributed/worker_state_machine.py [hunk 3]
```diff
@@ -1293,11 +1270,95 @@ def _ensure_communicating(self, *, stimulus_id: str) -> RecsInstrs:
...
+        3. if tied, by number of tasks available to be fetched (see note below)
+        4. if tied, by a random element
+
+        Omit workers that are either busy or in flight.
+        Remove peer workers with no tasks from data_needed.
+
+        Note
+        ----
+        Instead of number of tasks, we could've measured total nbytes and/or number of
+        tasks that only exist on the worker. Raw number of tasks is cruder but simpler.
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:1299]: Worth pointing out that, at this point, the previous algorithm was already random.

**Comment by crusaderky** [comment]
This PR is no longer blocked by #6593

File: distributed/worker_state_machine.py [hunk 4]
```diff
@@ -1292,11 +1269,95 @@ def _ensure_communicating(self, *, stimulus_id: str) -> RecsInstrs:
...
+        """
+        host = get_address_host(self.address)
+        heap = []
+
+        for worker, tasks in list(self.data_needed.items()):
+            if not tasks:
+                del self.data_needed[worker]
+                continue
+            if worker in self.in_flight_workers or worker in self.busy_workers:
+                continue
```
Comment (fjetter) [inline distributed/worker_state_machine.py:1308]: I know I initially suggested to go down this road but I would still like to raise the point that this implementation now would iterate over all workers every time we're calling `_ensure_communicating`

File: distributed/worker_state_machine.py [hunk 5]
```diff
@@ -1292,11 +1269,95 @@ def _ensure_communicating(self, *, stimulus_id: str) -> RecsInstrs:
...
+            # The number of tasks and possibly the top priority task may have changed
+            # since the last sort, since _select_keys_for_gather may have removed tasks
+            # that are also replicated on a higher-priority worker.
+            if not tasks:
+                del self.data_needed[worker]
+            elif -ntasks_neg != len(tasks):
+                heapq.heappush(
+                    heap,
+                    (is_remote, tasks.peek().priority, -len(tasks), rnd, worker, tasks),
+                )
```
Comment (fjetter) [inline distributed/worker_state_machine.py:1332]: This is tying a knot in my head when thinking about runtime.

File: distributed/worker_state_machine.py (hunk 4 above)
Comment (fjetter) [reply distributed/worker_state_machine.py:1308]: Even worse, if I'm not mistaken, the below iteration is not even linear (even without the additional push).

**Review by fjetter — COMMENTED** [review]
1. I'm slightly worried about runtimes for very large clusters (think ~1k workers upwards). Maybe this is not a big deal on the workers but I could see this amounting to a few orders of magnitude slower in the average case for large clusters
2. I think we should always prefer priority over anything else. Every other change should be verified with benchmarking
3. We had a similar argument when testing the Scheduler.rebalance algorithm in AMM but I'm again wondering if we should write dedicated unit tests for this, e.g. define these things as functions and test them in isolation


Testing could look like

```python

DATA_NEEDED = dict[str, HeapSet[TaskState]]

def _select_workers_for_gather(
    host: str,
    data_needed: DATA_NEEDED,
    skip_workers: set[str] # union of in_flight_workers and busy_workers
) -> Iterator[tuple[str, HeapSet[TaskState]]]:
    ...

def _select_keys_for_gather(
    data_needed,
    available,
) -> tuple[list[TaskState], int]:
    ...


def _get_data_to_fetch(host: str, data_needed, skip_workers)-> Iterator[tuple[str, set[TaskState], float]]:
    for worker, available in _select_keys_for_gather(
        host=host,
        data_needed=data_needed,
        skip_workers=skip_workers
    ):
        yield worker, *_select_keys_for_gather(data_needed=data_needed, available)



def test_prefer_priority():
    tasks = []
    data_needed = defaultdict(HeapSet)
    host = "127.0.0.1"
    local_worker = f"{host}:1234"
    remote_worker = "10.5.6.7:5678"

    ts1 = TaskState(
        key="key", 
        priority=1, 
        nbytes=1, 
        who_has={remote_worker}
    )
    data_needed[remote_worker].push(ts1)
    ts1 = TaskState(
        key="key2", 
        priority=0, 
        nbytes=1, 
        who_has={local_worker}
    )
    data_needed[local_worker].push(ts1)

    assert list(_get_data_to_fetch(host, data_needed, [])) == [
        (remote_worker, {ts1}, 1)
        (local_worker, {ts2}, 1)
    ]
```

This would at least provide us some lever to test a few of the edge cases encountered in this implementation. For instance, the mutability of data_needed and the cross play between `_select_keys_for_gather` and `_select_workers_for_gather`

File: distributed/worker_state_machine.py [hunk 6]
```diff
@@ -1292,11 +1269,95 @@ def _ensure_communicating(self, *, stimulus_id: str) -> RecsInstrs:
...
+        """Helper of _ensure_communicating.
+
+        Fetch all tasks that are replicated on the target worker within a single
+        message, up to target_message_size.
+        """
+        to_gather: list[TaskState] = []
+        total_nbytes = 0
+
+        while available:
+            ts = available.peek()
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/worker_state_machine.py:1350]: No. You don't pop if it would send you above the threshold.

File: distributed/worker_state_machine.py (hunk 4 above)
Comment (crusaderky) [reply distributed/worker_state_machine.py:1308]: The iteration is O(n) to the number of peer workers that hold any data to fetch, and O(n*logn) to the number of workers that hold data to fetch and are neither in flight nor busy - split between O(n) in python bytecode * O(logn) in C.

In a worst-case scenario of a cluster with 1000 workers where you suddenly have data to fetch from all 1000 (and I seriously doubt this is realistic), you'd perform 
1. a for loop in pure python of 1000 iterations
2. a single `heapq.heapify` on a list of 1000 elements, which costs O(n*logn) and is implemented in C
3. a for loop in pure python of 50 iterations (`distributed.worker.connections.outgoing`), each of which calls `heapq.heappop` on a list of 950~1000 elements, which costs O(logn) and is implemented in C

At the next call to _ensure_communicating, before any worker has responded, you'll just repeat step 1, while skipping steps 2 and 3.

I strongly suspect this whole thing to be negligible in most cases.

*I can easily improve this, moving aside the busy/in flight workers to a separate attribute (e.g. WorkerState.busy_workers), thus making the second call O(1).* However if you agree I'd rather do it in a follow-up PR.

File: distributed/worker_state_machine.py (hunk 5 above)
Comment (crusaderky) [reply distributed/worker_state_machine.py:1332]: heappop/heappush are O(logn) and are implemented in C

**Comment by crusaderky** [comment]
OK, I'm swapping priority and local/remote around.

> Well, the top priority task was never random but defined by the heap.

The previous algorithm was:
1. pick top priority task
2. given two tasks with the same priority, pick the one inserted first
3. of all hosts in ts.who_has, pick a random one among the local ones (see call to `random.choice`)
4. if no local peers exist, pick a random one among the remote ones
5. pick addition 50MB worth of lower-priority tasks from the same host, following priority first and insertion order next

> Do we have an option to not use random? How about a str-compare of tasks.peek() to make it deterministic?

That would be pointless when the same task is available in multiple replicas from different workers.
You could str-compare worker addresses, but that would be disastrous:
- when multiple workers are on the same host - lower-port workers would be hammered
- on a network where IP addresses convey locality - e.g. there's a ticket somewhere from a user that had their cluster spanning two AWS regions

I'm adding a statically seeded random state to obtain reproducibility.

File: distributed/worker_state_machine.py (hunk 5 above)
Comment (fjetter) [reply distributed/worker_state_machine.py:1332]: not talking about the heappush itself but rather that it is pushing on the heap we're currently iterating over.

File: distributed/worker_state_machine.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/worker_state_machine.py:1350]: I think what confuses me a lot here is that `available in self.data_needed.values()` is it not?
the way we shrink `available` is by calling remove when iterating over `ts.who_has`?

**Comment by crusaderky** [comment]
Would it be OK if I add unit tests (1) for the determinism and (2) to test that priority is chosen over locality as a follow-up after #6446? That PR makes it easier to write state tests.

**Comment by fjetter** [comment]
> Would it be OK if I add unit tests (1) [...] after https://github.com/dask/distributed/issues/6446?

sure

**Comment by crusaderky** [comment]
Added test about gather priority.
The performance test will be in https://github.com/dask/dask-benchmarks as discussed (PR to follow)

File: distributed/worker_state_machine.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/worker_state_machine.py:1366]: Yes. Added comment below.
